backend/cmm_helpers.ml

```diff
@@ -2721,101 +2721,108 @@ let send kind met obj args args_type result akind dbg =
+            dbg,
+            Any ) )
+  in
+  let loop_body =
+    ccatch
+      ( check_cont,
+        [VP.create check_li, typ_int; VP.create check_hi, typ_int],
+        dichotomy_expr,
+        check_expr,
+        dbg,
+        Any,
+        false )
+  in
   let li_vp = VP.create li in
   let hi_vp = VP.create hi in
-  let cont = Lambda.next_raise_count () in
-  let new_li = V.create_local "*new_li*" in
-  let new_hi = V.create_local "*new_hi*" in
-  let li_return = V.create_local "*li_return*" in
-  Clet
-    ( VP.create new_li,
-      ccatch
-        ( raise_num,
-          [VP.create li_return, typ_int],
-          Ccatch
-            ( Recursive,
-              [ ( cont,
-                  [li_vp, typ_int; hi_vp, typ_int],
-                  Clet
-                    ( VP.create mi,
-                      Cop
-                        ( Cor,
-                          [ Cop
-                              ( Clsr,
-                                [ Cop (Caddi, [Cvar li; Cvar hi], dbg);
-                                  cconst_int 1 ],
-                                dbg );
-                            cconst_int 1 ],
-                          dbg ),
-                      Cifthenelse
-                        ( Cop
-                            ( Ccmpi Clt,
-                              [ tag;
-                                Cop
-                                  ( mk_load_mut Word_int,
-                                    [ Cop
-                                        ( Cadda,
-                                          [ meths;
-                                            lsl_const (Cvar mi) log2_size_addr
-                                              dbg ],
-                                          dbg ) ],
-                                    dbg ) ],
-                              dbg ),
-                          dbg,
-                          Clet
-                            ( VP.create new_hi,
-                              Cop (Csubi, [Cvar mi; cconst_int 2], dbg),
-                              Cifthenelse
-                                (* here, [li] remains unchanged, but [hi] has
-                                   been replaced by [mi - 2] aka [new_hi]. *)
-                                ( Cop (Ccmpi Cge, [Cvar li; Cvar new_hi], dbg),
-                                  dbg,
-                                  Cexit (Lbl raise_num, [Cvar li], []),
-                                  dbg,
-                                  Cexit (Lbl cont, [Cvar li; Cvar new_hi], []),
-                                  dbg,
-                                  Any ) ),
-                          dbg,
-                          Cifthenelse
-                            (* here, we have replaced [li] by [mi]; [hi] remains
-                               unchanged. *)
-                            ( Cop (Ccmpi Cge, [Cvar mi; Cvar hi], dbg),
-                              dbg,
-                              Cexit (Lbl raise_num, [Cvar li], []),
```

Comment:
There is a one character patch that fixes the issue too, which is to replace `li` by `mi` on this line. I like my new code though, which I feel is a bit easier to understand (and closer to the original code in terms of the resulting assembly code).